From 97c33937a11f62756d785e7d121d4616e2340598 Mon Sep 17 00:00:00 2001
From: Aaron Williams <awilliams@marvell.com>
Date: Tue, 25 Feb 2020 22:20:48 -0800
Subject: [PATCH 497/767] mmc: octeontx2: fix handling calibration glitch

commit 2e2f7dd69ae3760cfbcef1fb93becbbc6b65f8ba from
git@git.assembla.com:cavium/WindRiver.linux.git

The logic was inverted.  This also increases the maximum speed for
CNF95XX and LOKI and adds another reset after calibration to match
U-Boot.

Change-Id: If51d04b89e9842490ad0263e73bb0504f6c0b5be
Signed-off-by: Aaron Williams <awilliams@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/23940
Reviewed-by: Chandrakala Chavva <cchavva@marvell.com>
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/mmc/host/cavium-thunderx.c | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/drivers/mmc/host/cavium-thunderx.c b/drivers/mmc/host/cavium-thunderx.c
index 3ee343c0bbb7..b1923ed2a357 100644
--- a/drivers/mmc/host/cavium-thunderx.c
+++ b/drivers/mmc/host/cavium-thunderx.c
@@ -72,7 +72,7 @@ static void thunder_calibrate_mmc(struct cvm_mmc_host *host)
 	if (host->cond_clock_glitch)
 		writeq(1, host->base + MIO_EMM_DEBUG(host));
 
-	if (!host->calibrate_glitch) {
+	if (host->calibrate_glitch) {
 		/*
 		 * Operation of up to 100 MHz may be achieved by skipping the
 		 * steps that establish the tap delays and instead assuming
@@ -138,6 +138,9 @@ static void thunder_calibrate_mmc(struct cvm_mmc_host *host)
 			how = "calibrated";
 		}
 
+		/* Reset eMMC subsystem */
+		writeq(0, host->base + MIO_EMM_CFG(host));
+		udelay(1);
 		/* restore old state */
 		writeq(emm_cfg, host->base + MIO_EMM_CFG(host));
 		mdelay(1);
@@ -160,7 +163,7 @@ static void thunder_calibrate_mmc(struct cvm_mmc_host *host)
 	 */
 	ps = (delay * PS_10000) / TOTAL_NO_OF_TAPS;
 	if (host->per_tap_delay != ps) {
-		dev_info(host->dev, "%s delay:%d per_tap_delay:%dpS\n",
+		dev_info(host->dev, "%s delay:%d per-tap delay:%dpS\n",
 			how, delay, ps);
 		host->per_tap_delay = ps;
 		host->delay_logged = 0;
@@ -254,10 +257,10 @@ static int thunder_mmc_probe(struct pci_dev *pdev,
 	case PCI_SUBSYS_DEVID_95XX:
 		if (rev == REV_ID_0)
 			host->cond_clock_glitch = true;
-		host->max_freq = MHZ_150;
+		host->max_freq = MHZ_167;
 		break;
 	case PCI_SUBSYS_DEVID_LOKI:
-		host->max_freq = MHZ_150;
+		host->max_freq = MHZ_167;
 		break;
 	default:
 		break;
-- 
2.31.1

